### PR TITLE
Samtools quickcheck

### LINKED ISSUE #9657: new module: samtools/quickcheck
implementing:
https://www.htslib.org/doc/samtools-quickcheck.html

### PR DESCRIPTION
**Pull request by Krannich479**
## PR checklist

Closes #9657  

- [x] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [x] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [ ] If necessary, include test data in your PR.
- [ ] Remove all TODO statements.
- [x] Broadcast software version numbers to `topic: versions` - [See version_topics](https://nf-co.re/blog/2025/version_topics)
- [x] Follow the naming conventions.
- [x] Follow the parameters requirements.
- [x] Follow the input/output options guidelines.
- [x] Add a resource `label`
- [ ] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - For modules:
    - [ ] `nf-core modules test <MODULE> --profile docker`
    - [ ] `nf-core modules test <MODULE> --profile singularity`
    - [x] `nf-core modules test <MODULE> --profile conda`
  - For subworkflows:
    - [ ] `nf-core subworkflows test <SUBWORKFLOW> --profile docker`
    - [ ] `nf-core subworkflows test <SUBWORKFLOW> --profile singularity`
    - [ ] `nf-core subworkflows test <SUBWORKFLOW> --profile conda`

## Summary
This PR adds `samtools quickcheck` to the nf-core modules. As opposed to many other tools, this samtools submodule does not return an output file. Instead, the command returns a non-zero exit code if the alignment file (SAM/BAM/CRAM) is invalid/malformatted.
The nf-core module specification does not strictly require output files. The idea here is to use the exit code of samtools for it's native behavior, i.e. the exit code determines the success of the nextflow workflow (regulated by the `errorStrategy`). If the input channel contains an invalid alignment file, the nf-core module returns a non-zero exit code and terminates the main worflow.

### CODE REVIEW

File: modules/nf-core/samtools/quickcheck/tests/main.nf.test [hunk 1]
```diff
@@ -0,0 +1,139 @@
+// TODO nf-core: Once you have added the required tests, please run the following command to build this file:
+// nf-core modules test samtools/quickcheck
```
Comment (mashehu) [inline modules/nf-core/samtools/quickcheck/tests/main.nf.test]:
```suggestion
```

File: modules/nf-core/samtools/quickcheck/tests/main.nf.test [hunk 2]
```diff
@@ -0,0 +1,139 @@
+// TODO nf-core: Once you have added the required tests, please run the following command to build this file:
```
Comment (prototaxites) [inline modules/nf-core/samtools/quickcheck/tests/main.nf.test]:
```suggestion
```

File: modules/nf-core/samtools/quickcheck/main.nf [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+    input:
+    tuple val(meta), path(bam)
+
+    output:
+    /*  There is not generated output files.
+        The sole purpose of 'samtools quickcheck' is to return an error if the mapping file is malformatted.
+        The purpose in of the Nextflow process is to break the main workflow (by default) if the script returns a non-zero exit code.
+        The nf-core components guidelines (https://nf-co.re/docs/guidelines/components/modules) do not specify that there must be output files.
+    */
```
Comment (prototaxites) [inline modules/nf-core/samtools/quickcheck/main.nf]: Suggestion: instead of erroring the process, we should capture the exitcode in the script and then return it.

This allows more flexibility to the module user - they can iterate over the outputs of the module and either filter out failed BAM files, throw an error, try and fix problems, etc.. - rather than just forcing the pipeline to die.

```suggestion
tuple val(meta), path(bam), env("EXIT_CODE")
```

File: modules/nf-core/samtools/quickcheck/main.nf [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+    when:
+    task.ext.when == null || task.ext.when
+
+    script:
+    def args = task.ext.args ?: ''
+    """
+    samtools quickcheck \\
+        $args \\
+        $bam
```
Comment (prototaxites) [inline modules/nf-core/samtools/quickcheck/main.nf]:
```suggestion
    samtools quickcheck \\
        $args \\
        $bam || EXIT_CODE=\$?
    EXIT_CODE=\${EXIT_CODE:-0}
```

File: modules/nf-core/samtools/quickcheck/main.nf [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    """
+
+    stub:
+    def args = task.ext.args ?: ''
+    """
+    echo $args
```
Comment (prototaxites) [inline modules/nf-core/samtools/quickcheck/main.nf:34]:
```suggestion
    EXIT_CODE=0
    echo $args
```

**Comment by Krannich479** [comment]
hey @prototaxites , thanks for looking into this, I am grateful for any feedback and suggestions. I do believe that's a good idea because a pipeline using the module with an exit status can redirect valid and invalid bam files into different channels afterwards.
Let me fix this :)
tbd

**Comment by Krannich479** [comment]
Much better functionality now imho. Adjusted meta.yml and tests accordingly. Linting still fails though.

File: modules/nf-core/samtools/quickcheck/meta.yml [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+          type: file
+          description: sequence_trace file
+          pattern: "*.{cram,sam,bam}"
+          ontologies:
+            - edam: http://edamontology.org/data_0924 # Sequence trace
+            - edam: http://edamontology.org/format_3462 # CRAM
+            - edam: http://edamontology.org/format_2573 # SAM
+            - edam: http://edamontology.org/format_2572 # BAM
+      - exit_code:
+          type: integer
```
Comment (prototaxites) [inline modules/nf-core/samtools/quickcheck/meta.yml]: My suspicion is that this technically will be returned as a string

File: modules/nf-core/samtools/quickcheck/tests/main.nf.test [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+                    file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/bam/test.paired_end.sorted.bam', checkIfExists: true)
+                ])
+                """
+            }
+        }
+
+        then {
+            assertAll (
+                { assert process.success },
+                { assert process.out.bam.collect{ it[2] } == ["0"] },
```
Comment (prototaxites) [inline modules/nf-core/samtools/quickcheck/tests/main.nf.test:28]: Would it be possible to find and add to test-datasets a broken BAM file? Just so we can have a test that captures a non-zero exit code also, to confirm that works.

Comment (prototaxites) [reply modules/nf-core/samtools/quickcheck/tests/main.nf.test:28]: https://github.com/samtools/samtools/tree/develop/test/quickcheck

Looks like there are some in the Samtools repo!

Comment (Krannich479) [reply modules/nf-core/samtools/quickcheck/tests/main.nf.test:28]: I did this in `test(" sam - PE - invalid file")` with a broken SAM file. I couldn't trivially convert this to a BAM file though since the file is invalid to samtools and `samtools view` cannot convert it. But the logic should remain the same.

File: modules/nf-core/samtools/quickcheck/meta.yml (hunk 6 above)
Comment (Krannich479) [reply modules/nf-core/samtools/quickcheck/meta.yml]: True. Will change description

File: modules/nf-core/samtools/quickcheck/tests/main.nf.test (hunk 7 above)
Comment (Krannich479) [reply modules/nf-core/samtools/quickcheck/tests/main.nf.test:28]: Is this what you had in mind?

File: modules/nf-core/samtools/quickcheck/meta.yml (hunk 6 above)
Comment (Krannich479) [reply modules/nf-core/samtools/quickcheck/meta.yml]: BTW, I got replies on the nf-core slack channel on this. The linting failed due to the multi-line comment in the output section.

File: modules/nf-core/samtools/quickcheck/tests/main.nf.test (hunk 7 above)
Comment (Krannich479) [reply modules/nf-core/samtools/quickcheck/tests/main.nf.test:28]: > https://github.com/samtools/samtools/tree/develop/test/quickcheck
> 
> Looks like there are some in the Samtools repo!

Nice catch. I'll try and include this!

Comment (prototaxites) [reply modules/nf-core/samtools/quickcheck/tests/main.nf.test:28]: Ah, I missed that - pesky `!=` hard to differentiate from `==` at a skim!

File: modules/nf-core/samtools/quickcheck/meta.yml (hunk 6 above)
Comment (prototaxites) [reply modules/nf-core/samtools/quickcheck/meta.yml]: Ah!!

File: modules/nf-core/samtools/quickcheck/tests/main.nf.test [hunk 8]
```diff
@@ -141,4 +142,64 @@ nextflow_process {
...
+                    file('https://github.com/samtools/samtools/raw/refs/heads/develop/test/quickcheck/1.quickcheck.badheader.bam', checkIfExists: true),
+                ]
+                """
+            }
+        }
+
+        then {
+            assertAll (
+                { assert !process.success }
+            )
```
Comment (prototaxites) [inline modules/nf-core/samtools/quickcheck/tests/main.nf.test:204]: Is there a reason this one fails instead of emitting an exitcode?

**Comment by Krannich479** [comment]
Added more negative test cases. The curious reader can now learn error codes from the test cases ;)

File: modules/nf-core/samtools/quickcheck/tests/main.nf.test (hunk 8 above)
Comment (Krannich479) [reply modules/nf-core/samtools/quickcheck/tests/main.nf.test]: Very reasonable question. I was curious about it too. I can only hypothesize why. Missing targets and missing EOF is detectable from the format specification and file structure, respectively. In contrast, a broken binary encoding in the header can result in virtually anything (erroneous) in plain SAM format. Maybe that's what makes samtools crash.

Comment (prototaxites) [reply modules/nf-core/samtools/quickcheck/tests/main.nf.test:204]: What is the error that occurs? I would have thought that `||` would capture any samtools failure and redirect the exit code, even if it doesn't return an "official" exit code

File: modules/nf-core/samtools/quickcheck/environment.yml [hunk 9]
```diff
@@ -0,0 +1,7 @@
+---
+# yaml-language-server: $schema=https://raw.githubusercontent.com/nf-core/modules/master/modules/environment-schema.json
+channels:
+  - conda-forge
+  - bioconda
+dependencies:
+  - "bioconda::samtools=1.23"
```
Comment (prototaxites) [inline modules/nf-core/samtools/quickcheck/environment.yml:7]:
```suggestion
  - "bioconda::samtools=1.23"
  - "bioconda::htslib=1.23"
```

It's been suggested to me before that it's important to match the htslib version to the samtools version in a conda env

File: modules/nf-core/samtools/quickcheck/tests/main.nf.test (hunk 8 above)
Comment (Krannich479) [reply modules/nf-core/samtools/quickcheck/tests/main.nf.test:204]: Double-checking... I might messed something up. brb

Comment (Krannich479) [reply modules/nf-core/samtools/quickcheck/tests/main.nf.test:204]: It was an oopsie on my side. Thanks for pointing it out again :)
